### PR TITLE
Fix crashes due to uninitialised fields

### PR DESCRIPTION
**Pull request by ffa-grahamdawes**
A number of controls in wxQT have empty default c'tors or don't check if their fields are NULL.  This commit fixes a number of these.

### CODE REVIEW

File: src/qt/dc.cpp [hunk 1]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
 void wxQtDCImpl::DoGetSize(int *width, int *height) const
 {
-    if (width)  *width  = m_qtPainter->device()->width();
-    if (height) *height = m_qtPainter->device()->height();
+    QPaintDevice *pDevice = m_qtPainter->device();
+
+    int deviceWidth;
+    int deviceHeight;
+
+    if (pDevice)
```
Comment (catalinr) [inline src/qt/dc.cpp]:
```suggestion
    if ( pDevice )
```

File: src/qt/dc.cpp [hunk 2]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
+    {
+        deviceWidth = pDevice->width();
+        deviceHeight = pDevice->height();
+    }
+    else
+    {
+        deviceWidth = 0;
+        deviceHeight = 0;
+
```
Comment (catalinr) [inline src/qt/dc.cpp]: This empty line should probably go after the closing `}`.

File: src/qt/dc.cpp [hunk 3]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
+    }
+    if (width) *width  = deviceWidth;
```
Comment (catalinr) [inline src/qt/dc.cpp]: As long as you're changing these lines, they could be brought in line with wxW coding style (which could include `{` `}` even for single line blocks).
```suggestion
    if ( width )
        *width = deviceWidth;
```

File: src/qt/dc.cpp [hunk 4]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
+    if (height) *height = deviceHeight;
```
Comment (catalinr) [inline src/qt/dc.cpp]:
```suggestion
    if ( height )
        *height = deviceHeight;
```

File: src/qt/dc.cpp [hunk 5]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
 void wxQtDCImpl::DoGetSizeMM(int* width, int* height) const
 {
-    if (width)  *width  = m_qtPainter->device()->widthMM();
-    if (height) *height = m_qtPainter->device()->heightMM();
+    QPaintDevice *pDevice = m_qtPainter->device();
+
+    int deviceWidthMM;
+    int deviceHeightMM;
+
+    if (pDevice)
```
Comment (catalinr) [inline src/qt/dc.cpp]:
```suggestion
    if ( pDevice )
```

File: src/qt/dc.cpp [hunk 6]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
+        deviceWidthMM = pDevice->widthMM();
+        deviceHeightMM = pDevice->heightMM();
+    }
+    else
+    {
+        deviceWidthMM = 0;
+        deviceHeightMM = 0;
+    }
+
+    if (width)  *width  = deviceWidthMM;
```
Comment (catalinr) [inline src/qt/dc.cpp]:
```suggestion
    if ( width )
        *width = deviceWidthMM;
```

File: src/qt/dc.cpp [hunk 7]
```diff
@@ -115,14 +115,46 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
+    if (height) *height = deviceHeightMM;
```
Comment (catalinr) [inline src/qt/dc.cpp]:
```suggestion
    if ( height )
        *height = deviceHeightMM;
```

**Review by catalinr — CHANGES_REQUESTED** [review]

File: src/qt/dc.cpp [hunk 8]
```diff
@@ -115,14 +115,50 @@ bool wxQtDCImpl::CanGetTextExtent() const
...
+    int deviceHeight;
+
+    if ( pDevice )
+    {
+        deviceWidth = pDevice->width();
+        deviceHeight = pDevice->height();
+    }
+    else
+    {
+        deviceWidth = 0;
```
Comment (vadz) [inline src/qt/dc.cpp]: I'm not sure it's a good idea to silently return (0,0) for an invalid DC. Is it really normal for a DC to be invalid in wxQt? This isn't the case in the other ports, AFAIK.

In which circumstances exactly does the crash due to the invalid `QPaintDevice` happen?

File: src/qt/radiobox.cpp [hunk 9]
```diff
@@ -149,7 +152,7 @@ bool wxRadioBox::Create(wxWindow *parent,
     // GetMajorDim() is the number of rows.
     if ( style & wxRA_SPECIFY_COLS )
         m_qtBoxLayout = new QHBoxLayout;
-    else if ( style & wxRA_SPECIFY_ROWS )
+    else
```
Comment (vadz) [inline src/qt/radiobox.cpp]: This doesn't seem to be related to the rest of the changes, probably merits its own commit.

File: src/qt/radiobox.cpp [hunk 10]
```diff
@@ -242,6 +245,9 @@ bool wxRadioBox::Show(unsigned int n, bool show)
 
 bool wxRadioBox::Show( bool show )
 {
+    if (!m_qtGroupBox)
```
Comment (vadz) [inline src/qt/radiobox.cpp]: This is not quite right, `Show()` can be called before creating the window, see `wxWindowMSW::Show()` in `src/msw/window.cpp`. So we should call the base class version first instead of always returning `false` in this case.

File: src/qt/radiobox.cpp (hunk 9 above)
Comment (ffa-grahamdawes) [reply src/qt/radiobox.cpp]: Is there an easy way to split the commit?  I lumped it together as without the an unconditional else m_qtBoxLayout can be left uninitialised,  but I agree that it isn't consistent with the rest of the commit.

File: src/qt/dc.cpp (hunk 8 above)
Comment (ffa-grahamdawes) [reply src/qt/dc.cpp]: In wxQT if you use a memory DC without first calling SelectObject,  or by calling SelectObject with a "null" bitmap.  Then the QPainter will have a NULL QPaintDevice.

File: src/qt/radiobox.cpp (hunk 9 above)
Comment (vadz) [reply src/qt/radiobox.cpp]: Well, it's easy to do with `git rebase --interactive`, but only if you're used to using it (basically you need to "edit" this commit in the interactive rebase TODO list and then separately commit this part, with its own commit message, and the rest). Alternatively, you could amend this commit to undo this change manually and then make a separate commit with it.

File: src/qt/dc.cpp (hunk 8 above)
Comment (vadz) [reply src/qt/dc.cpp]: Could it be better to do this check in `wxMemoryDCImpl` only? For it we clearly do need it, as in the situation you describe `wxMemoryDC::GetSize()` returns (0,0) under the other platforms too, but for other kinds of DC I'm not so sure...

Oh well, I guess we can always do it later.

**Comment by ffa-grahamdawes** [comment]
Would it be possible to apply all the changes except the radiobox.cpp change please?.  I'll then make a new PR with just that change.

Thanks
